dev-docs/source/MVPTutorial/CompleteGUI.rst

```diff
@@ -168,12 +200,119 @@ Presenter
+
+        def __init__(self, parent=None):
+            super(view, self).__init__(parent)
+
+            grid = QtWidgets.QVBoxLayout(self)
+
+            self.table = QtWidgets.QTableWidget(self)
+            self.table.setRowCount(4)
+            self.table.setColumnCount(2)
+
+            grid.addWidget(self.table)
+
+            self.colours = QtWidgets.QComboBox()
+            options=["Blue", "Green", "Red"]
+            self.colours.addItems(options)
+
+            self.grid_lines= QtWidgets.QTableWidgetItem()
+            self.grid_lines.setFlags(QtCore.Qt.ItemIsUserCheckable | QtCore.Qt.ItemIsEnabled)
+            self.grid_lines.setCheckState(QtCore.Qt.Unchecked)
+            self.addItemToTable("Show grid lines", self.grid_lines, 1)
+
+            self.freq = QtWidgets.QTableWidgetItem("1.0")
+            self.phi = QtWidgets.QTableWidgetItem("0.0")
+
+            self.addWidgetToTable("Colour", self.colours, 0)
+            self.addItemToTable("Frequency", self.freq, 2)
+            self.addItemToTable("Phase", self.phi, 3)
+
+            self.plot = QtWidgets.QPushButton('Add', self)
+            self.plot.setStyleSheet("background-color:lightgrey")
+
+            grid.addWidget(self.plot)
+
+            self.setLayout(grid)
+
+            self.plot.clicked.connect(self.buttonPressed)
+
+        def getColour(self):
+            return self.colours.currentText()
+
+        def getGridLines(self):
+            return self.grid_lines.checkState() == QtCore.Qt.Checked
+
+        def getFreq(self):
+            return float(self.freq.text())
+
+        def getPhase(self):
+            return float(self.phi.text())
+
+        def buttonPressed(self):
+            self.plotSignal.emit()
+
+        def setTableRow(self, name, row):
+            text = QtWidgets.QTableWidgetItem(name)
+            text.setFlags(QtCore.Qt.ItemIsEnabled)
+            col = 0
+            self.table.setItem(row, col, text)
+
+        def addWidgetToTable(self, name, widget, row):
+            self.setTableRow(name, row)
+            col = 1
+            self.table.setCellWidget(row, col, widget)
+
+        def addItemToTable(self, name, widget, row):
+            self.setTableRow(name, row)
+            col = 1
+            self.table.setItem(row, col, widget)
+
+        def setColours(self, options):
+            self.colours.clear()
+            self.colours.addItems(options)
+
+``model.py``
+############
+
+.. code-block:: python
+
+    from __future__ import (absolute_import, division, print_function)
+    import numpy as np
+
+    class DataGenerator(object):
+
+        def __init__(self):
+            self.x_data = np.linspace(0.0, 10.0, 100)
+            self.y_data = []
+
+        def genData(self, freq, phi):
+            self.y_data = np.sin(freq * self.x_data + phi)
+
+        def getXData(self):
+            return self.x_data
+
+        def getYData(self):
+            return self.y_data
+
+
+    class ColourConverter(object):
```

Comment:
I know this is a toy/demo implementation but this is a good example of a bad practice we see all too often in Mantid. This could easily be a single function and gains no benefit from being a class.

Since there is already part of the model that uses a class (correctly) I'd rather this was converted to be a function for the sake of showing this is an acceptable (and encouraged) thing to do when appropriate.

Reply:
Would it be more acceptable to split it up unto two scripts, a ``model_colour.py`` and a ``model_data.py``, in the same way as the presenter and view are split?

Reply:
That's a separate concern and IMO is not worth it for this demo (having too many files would make it more difficult to follow). The main concern here is class vs free functions (and avoiding classes when they don't contribute anything to functionality).

Reply:
This is me thinking based on the explination from the previous version of this tutorial in ``model.rst``

> In this case we have a distinct second method. Usually this will be placed into its own file, however for simplicity we will contain it within the same file as the above code.

suggesting that since generating a line from an function and giving a list of line colours are distinct tasks they should be kepts seperatly.

Reply:
Yes it would be more correct. It's probably a good idea to make that change too.

Reply:
Ok so I've seperated them into two model scripts, for simplicity I've not linked them into a master model, instead I've just just imported them both into ``main``

Reply:
`ColorModel` still needs simplifying. Specifically it isn't even a model as it holds no state. It can easily be a free function or even a global constant.

Reply:
What do you mean by state?

Reply:
No member variables that are modified. That and the fact the class essentially wraps an existing container type mean it should not be a class at all.

It is good practice to avoid classes where possible, this has the benefit of easier reuse and testing.

Reply:
Ok so now listing and returning the line colours have been made into a function